src/murmur/ServerDB.cpp

```diff
@@ -2156,26 +2156,55 @@ int Server::readLastChannel(int id) {
 	TransactionHolder th;
 	QSqlQuery &query = *th.qsqQuery;
 
-	SQLPREP("SELECT `lastchannel`,`last_disconnect` FROM `%1users` WHERE `server_id` = ? AND `user_id` = ?");
+	SQLPREP("SELECT `lastchannel`,`last_active`,`last_disconnect` FROM `%1users` WHERE `server_id` = ? AND `user_id` = ?");
 	query.addBindValue(iServerNum);
 	query.addBindValue(id);
 	SQLEXEC();
 
 	if (query.next()) {
 		int cid = query.value(0).toInt();
-		if (query.value(1).isNull()) {
-			return qhChannels.contains(cid) ? cid : -1;
-		}
-		QDateTime last_disconnect = QDateTime::fromString(query.value(1).toString(), Qt::ISODate);
-		last_disconnect.setTimeSpec(Qt::UTC);
-		QDateTime now = QDateTime::currentDateTime();
-		now           = now.toTimeSpec(Qt::UTC);
+
+		if (!qhChannels.contains(cid))
+			return -1;
 
 		int duration = Meta::mp.iRememberChanDuration;
-		if (duration <= 0 || last_disconnect.secsTo(now) <= duration) {
-			if (qhChannels.contains(cid))
-				return cid;
+
+		if (duration <= 0)
+			return cid;
+
+		QDateTime last_active, last_disconnect;
+
+		// NULL columns for last_active and last_disconnect will create empty invalid QDateTime objects.
+		// Using such objects with QDateTime::secsTo() will return 0 as per Qt specification.
+
+		if (!query.value(1).isNull()) {
+			last_active = QDateTime::fromString(query.value(1).toString(), Qt::ISODate);
+			last_active.setTimeSpec(Qt::UTC);
+		} else {
+			last_active = QDateTime();
+		}
+
+		if (!query.value(2).isNull()) {
+			last_disconnect = QDateTime::fromString(query.value(2).toString(), Qt::ISODate);
+			last_disconnect.setTimeSpec(Qt::UTC);
+		} else {
+			last_disconnect = QDateTime();
 		}
+
+		if (last_active.secsTo(last_disconnect) <= 0) {
+			// Calculating the duration from last_disconnect is unreliable due to invalid values.
+			// Use the server uptime as time reference instead.
+			last_disconnect = QDateTime::currentDateTime();
+			last_disconnect.setTimeSpec(Qt::UTC);
+			qint64 uptimeMSecs = static_cast< qint64 >(tUptime.elapsed() / 1000ULL);
+			last_disconnect = last_disconnect.addMSecs(-uptimeMSecs);
+		}
+
+		QDateTime now = QDateTime::currentDateTime();
+		now.setTimeSpec(Qt::UTC);
+
+		if (last_disconnect.secsTo(now) <= duration)
+			return cid;
```

Comment:
```suggestion
		if (last_disconnect.secsTo(now) <= duration) {
			return cid;
		}
```

Reply:
Because it visually encloses the if-body, making the code easier to read and if in the future someone wants to add something to that body, they don't have to add the braces first. The emphasis lies on the first argument though.